src/main/java/org/springframework/data/elasticsearch/Document.java

```diff
@@ -0,0 +1,388 @@
+	 *
+	 * @param map source map containing key-value pairs and sub-documents.
+	 * @return a new {@link Document}.
+	 */
+	static Document from(Map<String, Object> map) {
+
+		Assert.notNull(map, "Map must not be null");
+
+		if (map instanceof LinkedHashMap) {
+			return new MapDocument(map);
+		}
+
+		return new MapDocument(new LinkedHashMap<>(map));
+	}
+
+	/**
+	 * Parse JSON to {@link Document}.
+	 *
+	 * @param json must not be {@literal null}.
+	 * @return the parsed {@link Document}.
+	 */
+	static Document parse(String json) {
+		try {
+			return new MapDocument(MapDocument.OBJECT_MAPPER.readerFor(Map.class).readValue(json));
+		} catch (IOException e) {
+			throw new ElasticsearchException("Cannot parse JSON", e);
+		}
+	}
+
+	/**
+	 * {@link #put(Object, Object)} the {@code key}/{@code value} tuple and return {@code this} {@link Document}.
+	 *
+	 * @param key key with which the specified value is to be associated.
+	 * @param value value to be associated with the specified key.
+	 * @return {@code this} {@link Document}.
+	 */
+	default Document append(String key, Object value) {
+		put(key, value);
+		return this;
+	}
+
+	/**
+	 * Return {@literal true} if this {@link Document} is associated with an identifier.
+	 *
+	 * @return {@literal true} if this {@link Document} is associated with an identifier, {@literal false} otherwise.
+	 */
+	default boolean hasId() {
+		return false;
+	}
+
+	/**
+	 * Retrieve the identifier associated with this {@link Document}.
+	 * <p>
+	 * The default implementation throws {@link UnsupportedOperationException}. It's recommended to check {@link #hasId()}
+	 * prior to calling this method.
+	 *
+	 * @return the identifier associated with this {@link Document}.
+	 * @throws IllegalStateException if the underlying implementation supports Id's but no Id was yet associated with the
+	 *           document.
+	 */
+	default String getId() {
+		throw new UnsupportedOperationException();
+	}
+
+	/**
+	 * Set the identifier for this {@link Document}.
+	 * <p>
+	 * The default implementation throws {@link UnsupportedOperationException}.
+	 */
+	default void setId(String id) {
+		throw new UnsupportedOperationException();
+	}
+
+	/**
+	 * Return {@literal true} if this {@link Document} is associated with a version.
+	 *
+	 * @return {@literal true} if this {@link Document} is associated with a version, {@literal false} otherwise.
+	 */
+	default boolean hasVersion() {
+		return false;
+	}
+
+	/**
+	 * Retrieve the version associated with this {@link Document}.
+	 * <p>
+	 * The default implementation throws {@link UnsupportedOperationException}. It's recommended to check
+	 * {@link #hasVersion()} prior to calling this method.
+	 *
+	 * @return the version associated with this {@link Document}.
+	 * @throws IllegalStateException if the underlying implementation supports Id's but no Id was yet associated with the
+	 *           document.
+	 */
+	default long getVersion() {
```

Comment:
Because we want to reduce nullability. If a version is associated with an object, then we want to retrieve it. If there's no version, then we get an exception. Therefore, we expose `has…()` methods. It is a strict approach but eliminates the probability of `NullPointerException` downstream. Similar approach with `getBooleanOrDefault`, `getLongOrDefault`.